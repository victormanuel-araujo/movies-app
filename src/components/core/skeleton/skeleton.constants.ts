export const SKELETON_GRADIENT_HEX_OPACITY = '22';
