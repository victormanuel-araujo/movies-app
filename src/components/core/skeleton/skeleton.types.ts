import { ThemeColorNames } from '$config/theme/theme.types';

export interface SkeletonProps {
  height: number | string;
  width: number | string;
  radius?: number;

  color: ThemeColorNames;
  backgroundColor: ThemeColorNames;
}
