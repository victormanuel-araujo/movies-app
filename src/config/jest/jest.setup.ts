import '@testing-library/jest-native/extend-expect';

jest.setTimeout(10000);
