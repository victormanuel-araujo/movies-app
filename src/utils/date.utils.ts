const WEEK_DAYS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
const MONTHS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

export function toDateString(date?: string): string {
  if (!date) return '';
  const dateObj = new Date(date);

  return `${WEEK_DAYS[dateObj.getDay()]}, ${dateObj.getDate()} ${
    MONTHS[dateObj.getMonth()]
  } ${dateObj.getFullYear()}`;
}

export function minutesToHour(minutes?: number) {
  if (!minutes) return '';

  const remainMinutes = minutes % 60;
  const hours = (minutes - remainMinutes) / 60;
  return hours > 0 ? `${hours}h${remainMinutes}min` : `${remainMinutes}min`;
}
