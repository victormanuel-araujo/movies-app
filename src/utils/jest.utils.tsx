import { render } from '@testing-library/react-native';

import { theme } from '$config/theme';
import React from 'react';
import { ThemeProvider } from 'styled-components/native';

export function renderWithTheme(children: React.ReactNode) {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
}
