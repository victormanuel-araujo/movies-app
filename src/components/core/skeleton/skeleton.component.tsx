import { Animated } from 'react-native';

import { SKELETON_GRADIENT_HEX_OPACITY } from './skeleton.constants';
import * as Styles from './skeleton.styles';
import { SkeletonProps } from './skeleton.types';

import { getThemeColor } from '$config/theme/theme.utils';
import React, { useEffect } from 'react';
import { useTheme } from 'styled-components/native';

export const Skeleton = ({ color, ...props }: SkeletonProps) => {
  const theme = useTheme();
  const [animation] = React.useState(new Animated.Value(0));

  function getGradientColors() {
    const hexColor = getThemeColor(color)({ theme });
    const hexColorWithOpacity = `${hexColor}${SKELETON_GRADIENT_HEX_OPACITY}`;

    return [hexColorWithOpacity, hexColor, hexColorWithOpacity];
  }

  useEffect(() => {
    const animationLoop = Animated.loop(
      Animated.sequence([
        Animated.timing(animation, {
          duration: 1500,
          useNativeDriver: false,
          toValue: 1,
        }),
        Animated.timing(animation, { duration: 0, useNativeDriver: false, toValue: 0 }),
      ]),
    );

    animationLoop.start();

    return animationLoop.stop;
  }, []);

  const animationInterpolate = animation.interpolate({
    inputRange: [0, 1],
    outputRange: ['-100%', '120%'],
  });

  return (
    <Styles.SkeletonWrapper {...props} testID="skeleton-wrapper">
      <Styles.SkeletonGradient
        end={{ y: 0, x: 1 }}
        start={{ y: 0, x: 0 }}
        testID="skeleton-gradient"
        colors={getGradientColors()}
        style={{ left: animationInterpolate }}
      />
    </Styles.SkeletonWrapper>
  );
};
