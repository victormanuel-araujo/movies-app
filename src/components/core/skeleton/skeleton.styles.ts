import { Animated } from 'react-native';
import LinearGradient from 'react-native-linear-gradient';

import { SkeletonProps } from '@components/core/skeleton/skeleton.types';

import { getThemeColor } from '$config/theme/theme.utils';
import styled, { css } from 'styled-components/native';

const AnimatedLinearGradient = Animated.createAnimatedComponent(LinearGradient);

export const SkeletonWrapper = styled.View<
  Pick<SkeletonProps, 'backgroundColor' | 'height' | 'width' | 'radius'>
>(
  ({ backgroundColor, height, width, radius, theme }) => css`
    overflow: hidden;
    position: relative;
    border-radius: ${radius}px;
    height: ${typeof height === 'number' ? `${height}px` : height};
    width: ${typeof width === 'number' ? `${width}px` : width};
    background-color: ${getThemeColor(backgroundColor)({ theme })};
  `,
);

export const SkeletonGradient = styled(AnimatedLinearGradient)`
  height: 100%;
  width: 60px;
  position: absolute;
  top: 0px;
  left: 0px;
`;
