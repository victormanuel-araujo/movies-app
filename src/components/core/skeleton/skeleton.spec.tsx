import { View } from 'react-native';

import { Skeleton } from '@components/core/skeleton/skeleton.component';

import { renderWithTheme } from '#utils/jest.utils';

import React from 'react';

describe('Skeleton component', () => {
  it('should match snapshot', async () => {
    const component = renderWithTheme(
      <Skeleton height={20} width={100} color="primary.base" backgroundColor="primary.light" />,
    );

    expect(component).toMatchSnapshot();
  });

  it('should set the gradient colors correctly', async () => {
    const component = renderWithTheme(
      <Skeleton height={20} width={100} color="primary.base" backgroundColor="primary.light" />,
    );

    const skeletonGradient = await component.findByTestId('skeleton-gradient');

    expect(skeletonGradient.props.colors).toHaveLength(3);
  });

  it('should set height and width params correctly', async () => {
    const componentWithStringParams = await renderWithTheme(
      <View style={{ height: 20, width: 100 }}>
        <Skeleton height="50%" width="100%" color="primary.base" backgroundColor="primary.light" />,
      </View>,
    ).findByTestId('skeleton-wrapper');

    const componentWithNumberParams = await renderWithTheme(
      <Skeleton height={20} width={100} color="primary.base" backgroundColor="primary.light" />,
    ).findByTestId('skeleton-wrapper');

    expect(componentWithStringParams).toHaveStyle({ height: '50%', width: '100%' });
    expect(componentWithNumberParams).toHaveStyle({ height: 20, width: 100 });
  });
});

// 580142048
