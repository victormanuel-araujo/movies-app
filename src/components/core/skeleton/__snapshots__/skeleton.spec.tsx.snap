// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton component should match snapshot 1`] = `
<View
  backgroundColor="primary.light"
  height={20}
  style={
    [
      {
        "backgroundColor": "#B98AE6",
        "borderRadius": "px",
        "height": 20,
        "overflow": "hidden",
        "position": "relative",
        "width": 100,
      },
    ]
  }
  testID="skeleton-wrapper"
  width={100}
>
  <BVLinearGradient
    collapsable={false}
    colors={
      [
        580142048,
        4287906784,
        580142048,
      ]
    }
    endPoint={
      {
        "x": 1,
        "y": 0,
      }
    }
    locations={null}
    startPoint={
      {
        "x": 0,
        "y": 0,
      }
    }
    style={
      {
        "height": "100%",
        "left": "-100%",
        "position": "absolute",
        "top": 0,
        "width": 120,
      }
    }
    testID="skeleton-gradient"
  />
</View>
`;
